### PR TITLE
Always create a temporary for iterated expressions in a for-of loop

### LINKED ISSUE #5475: Wrong ES3/ES5 code when iterated identifier is reassigned in a 'for...of' statement
There is an inconsistency regarding the emitted ES3/ES5 code:

```` TypeScript
var x = [3, 3, 3];
for (var i of x)
{
	x = [2, 2];
	console.log(i);
}
````

compiles to 

```` JavaScript
var x = [3, 3, 3];
for (var _i = 0; _i < x.length; _i++) {
    var i = x[_i];
    x = [2, 2];
    console.log(i);
}
````

printing

    3
    2

instead of (ES6 semantics)

    3
    3
    3

(Note that the code emitted also differs from the feature description: https://github.com/Microsoft/TypeScript/pull/2207#issue-59900754)

---
On the other hand, merely adding parentheses around the expression:

```` TypeScript
var x = [3, 3, 3];
for (var i of (x))
{
    x = [2, 2];
    console.log(i);
}
````

**correctly** compiles to

```` JavaScript
var x = [3, 3, 3];
for (var _a = 0, _b = (x); _a < _b.length; _a++) {
    var i = _b[_a];
    x = [2, 2];
    console.log(i);
}
````

giving the expected output.

---
In other words, it looks like there is some sort of faulty optimization applied: As soon as the expression to iterate over is a bare variable, it is not rescued to a new local (here: `_b`). As shown, this results in wrong semantics.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #5475. The only special handling we use now is that we create a name based on the original identifier if we had an identifier.

### CODE REVIEW

**Comment by mhegazy** [comment]
@rbuckton can you take a look.

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -3614,10 +3614,12 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                 //
                 //     for (let v of arr) { }
                 //
-                // we don't want to emit a temporary variable for the RHS, just use it directly.
-                let rhsIsIdentifier = node.expression.kind === SyntaxKind.Identifier;
+                // we can't reuse 'arr' because it might be modified within the body of the loop.
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: One possibility might be to check if the rhs is declared as `const`, and then we can emit it inline.

**Comment by rbuckton** [comment]
Other than the `const` comment, :+1:

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -3614,10 +3614,12 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
                 //
-                // we don't want to emit a temporary variable for the RHS, just use it directly.
-                let rhsIsIdentifier = node.expression.kind === SyntaxKind.Identifier;
+                // we can't reuse 'arr' because it might be modified within the body of the loop.
                 let counter = createTempVariable(TempFlags._i);
-                let rhsReference = rhsIsIdentifier ? <Identifier>node.expression : createTempVariable(TempFlags.Auto);
+                let rhsReference = createSynthesizedNode(SyntaxKind.Identifier) as Identifier;
+                rhsReference.text = node.expression.kind === SyntaxKind.Identifier ?
+                    makeUniqueName((<Identifier>node.expression).text) :
+                    makeTempVariableName(TempFlags.Auto);
```
Comment (yuit) [inline src/compiler/emitter.ts]: Could you add tests for this case?

**Comment by yuit** [comment]
otherwise lgtm

File: src/compiler/emitter.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: We already have tests for that case

**Pull request merged**